### PR TITLE
STM32F4 definitions for flash controller

### PR DESCRIPTION
**Pull request by ag-123**
i've created the following files:
STM32F4/cores/maple/libmaple/flashF4.h
STM32F4/cores/maple/libmaple/flashF4.c
    
intended to replace
STM32F4/cores/maple/libmaple/flash.h
STM32F4/cores/maple/libmaple/flash.c

note that while the file does not change the contents of any files, i.e. they are totally new files.

but merging this will possibly cause duplicated function name conflicts
``void flash_set_latency(uint32 wait_states);``
has been redefined but has same name

``void flash_enable_prefetch(void);``
becomes
```
void flash_enable_ART(void);
void flash_disable_ART(void); //this is new
```
to merge this, it is (likely) necessary to remove flash.h and flash.c, you may like to keep a backup of those.
then for files that ``#include <flash.h>`` e.g. rccF4.c
it would be necessary to point to the new include flashF4.h instead

note that i did not do any edits in rccF4.c or anywhere else. i just presented 
flashF4.h and flashF4.c. this is to prevent possible merge conflicts

i did some basic tests for the functions 
flash_set_latency()
and 
flash_enable_ART() - this basically enable prefetch + instruction cache + data cache
flash_disable_ART() - this disable prefetch + instruction cache + data cache
they worked ok

the definitions are taken from the official core includes for stm32f4xx
i cross referenced f401, f411, (f405, f407, f415, f417 rm0009) and copied/adapted the definitions
pretty much all of them are there, i merged the definitions so long as they are non conflicting

there are no ifdefs, i place comments for entries that should not be used for particular mcus ;)

### CODE REVIEW

**Comment by stevstrong** [comment]
It would be very nice if you could make the changes in the existing files, and not add any new ones.
The F4 index makes no point because all the files are F4 related.
I see that this involves removal of some defines from the RCC module.

**Comment by ag-123** [comment]
i did some tests
i deleted and replace flash.h and flash.c with flashF4.h and flashF4.c
they actually build and works without errors.
just with warnings about the symbols being redefined.
they are literally the same definitions.

do i go about updating rccF4.c ?

File: STM32F4/cores/maple/libmaple/flash.h [hunk 1]
```diff
@@ -64,77 +65,167 @@ typedef struct flash_reg_map {
...
+
+/* Flash KEYR */
+
+#define FLASH_KEYR_KEY1				   0x45670123
+#define FLASH_KEYR_KEY2				   0xCDEF89AB
+
+/* FLASH_OPTKEYR */
+
+#define FLASH_OPTKEYR_KEY1			   0x08192A3B
+#define FLASH_OPTKEYR_KEY2			   0x4C5D6E7F
```
Comment (stevstrong) [inline STM32F4/cores/maple/libmaple/flash.h]: Please review indentation, al spaces (preferably) or all tabs.

File: STM32F4/cores/maple/libmaple/flash.h [hunk 2]
```diff
@@ -64,77 +65,167 @@ typedef struct flash_reg_map {
...
+//SNB - sector number
+//note: for F401, F405, F411, F417 this is 4 bits, F42xxx, F43xxx 5 bits
+//for 4 bits one, 5th bit (i.e. absolute bit 7) must be 0
+#define FLASH_CR_SNB_Msk               BIT_MASK_SHIFT(0x1FUL, FLASH_CR_SNB_Pos)
+#define FLASH_CR_PSIZE_Pos             (8U)
+#define FLASH_CR_PSIZE_Msk             BIT_MASK_SHIFT(0x3UL, FLASH_CR_PSIZE_Pos)
+#define FLASH_CR_PSIZE8X			   BIT_MASK_SHIFT(0, FLASH_CR_PSIZE_Pos)
+#define FLASH_CR_PSIZE16X			   BIT_MASK_SHIFT(1U, FLASH_CR_PSIZE_Pos)
+#define FLASH_CR_PSIZE32X			   BIT_MASK_SHIFT(2U, FLASH_CR_PSIZE_Pos)
+#define FLASH_CR_PSIZE64X			   BIT_MASK_SHIFT(3U, FLASH_CR_PSIZE_Pos)
```
Comment (stevstrong) [inline STM32F4/cores/maple/libmaple/flash.h]: Please review indentation.

**Comment by ag-123** [comment]
strange, the indents looks just well in eclipse
updating ...

**Comment by stevstrong** [comment]
Thanks.

**Pull request merged**